src/stateful-event-subscriber.ts

```diff
@@ -20,13 +27,85 @@ export abstract class StatefulEventSubscriber<State>
   protected invalid: boolean = false;
 
   isTracking: () => boolean = () => false;
+  public addressesSubscribed: string[] = [];
+
+  private mapKey: string;
+  private cacheName: string;
 
-  constructor(public readonly name: string, protected logger: Logger) {}
+  public name: string;
+
+  constructor(
+    public readonly parentName: string,
+    _name: string,
+    protected dexHelper: IDexHelper,
+    protected logger: Logger,
+    private masterPoolNeeded: boolean = false,
+  ) {
+    this.name = _name.toLowerCase();
+    this.mapKey =
+      `${CACHE_PREFIX}_${this.dexHelper.config.data.network}_${this.parentName}`.toLowerCase();
+    this.cacheName = `${this.mapKey}_${this.name}`.toLowerCase();
+  }
 
   getStateBlockNumber(): Readonly<number> {
     return this.stateBlockNumber;
   }
 
+  async initialize(
+    blockNumber: number,
+    cb?: (state: DeepReadonly<State>) => void,
+  ) {
+    if (this.dexHelper.config.isSlave && this.masterPoolNeeded) {
+      let stateAsString = await this.dexHelper.cache.hget(
+        this.mapKey,
+        this.name,
+      );
+
+      // if there is a state in cache
+      if (stateAsString) {
+        const state: StateCache<State> = Utils.Parse(stateAsString);
+
+        if (state.state === null) {
+          this.logger.warn(
+            `${this.parentName}: ${this.name}: found null state in cache generate new one`,
+          );
+          state.state = await this.generateState(blockNumber);
+        }
+        // apply a callback on the state (usefull for initialization for slaves instance)
+        if (cb) {
+          cb(state.state);
+        }
+        this.logger.debug(
+          `${this.parentName}: ${this.name}: found state from cache`,
+        );
+        // set state and the according blockNumber. state.bn can be smaller, greater or equal
+        // to blockNumber
+        this.setState(state.state, state.bn);
```

Comment:
`state.bn` doesn't seem to be necessarily the block number, if the state was generated locally using `blockNumber`.